core/templates/pages/topic-editor-page/services/topic-editor-state.service.ts

```diff
@@ -18,528 +18,544 @@
-              topicBackendObject.skillIdToDescriptionDict
-            );
-            _updateSkillIdToRubricsObject(
-              topicBackendObject.skillIdToRubricsDict);
-            var changeList = UndoRedoService.getCommittableChangeList();
-            for (var i = 0; i < changeList.length; i++) {
-              if (changeList[i].cmd === 'delete_canonical_story' ||
-                  changeList[i].cmd === 'delete_additional_story') {
-                EditableStoryBackendApiService.deleteStory(
-                  changeList[i].story_id);
-              }
-            }
-            UndoRedoService.clearChanges();
-            _topicIsBeingSaved = false;
-            if (successCallback) {
-              successCallback();
+      // Don't attempt to save the topic if there are no changes pending.
+      if (!this.undoRedoService.hasChanges()) {
+        reject();
+      }
+      this._topicIsBeingSaved = true;
+      this.editableTopicBackendApiService.updateTopic(
+        this._topic.getId(), this._topic.getVersion(),
+        commitMessage, this.undoRedoService.getCommittableChangeList()).then(
+        (topicBackendObject) => {
+          this._updateTopic(
+            topicBackendObject.topicDict,
+            topicBackendObject.skillIdToDescriptionDict
+          );
+          this._updateSkillIdToRubricsObject(
+            topicBackendObject.skillIdToRubricsDict);
+          let changeList =
+         this.undoRedoService.getCommittableChangeList();
+          for (let change of changeList) {
+            if ((change.cmd === 'delete_canonical_story' ||
+                  change.cmd === 'delete_additional_story') &&
+                  'story_id' in change) {
+              this.editableStoryBackendApiService.deleteStory(
+                change.story_id);
             }
-            $rootScope.$applyAsync();
-          }, function(error) {
-            AlertsService.addWarning(
-              error || 'There was an error when saving the topic.');
-            _topicIsBeingSaved = false;
-          });
-        return true;
-      },
+          }
+          this.undoRedoService.clearChanges();
+          this._topicIsBeingSaved = false;
+          resolve();
+        }, (error: string) => {
+          this.alertService.addWarning(
+            error || 'There was an error when saving the topic.');
+          this._topicIsBeingSaved = false;
+        });
+    });
+  }
 
-      /**
-       * Returns whether this service is currently attempting to save the
-       * topic maintained by this service.
-       */
-      isSavingTopic: function() {
-        return _topicIsBeingSaved;
-      },
+  /**
+   * Returns whether this service is currently attempting to save the
+   * topic maintained by this service.
+   */
+  isSavingTopic(): boolean {
+    return this._topicIsBeingSaved;
+  }
 
-      get onTopicInitialized() {
-        return _topicInitializedEventEmitter;
-      },
+  get onTopicInitialized(): EventEmitter<unknown> {
```

Comment:
Do similar changes in this file as well.
```suggestion
  get onTopicInitialized(): EventEmitter<void> {
```

Reply:
Done